### PR TITLE
Improve/unify error handling, plumb through compilation errors to the browser console

### PR DESCRIPTION
**Pull request by jakemac53**
Added some new Exception types which are thrown instead of AssetNotFoundExceptions, and added uniform handling for them which creates a new asset at `originalId.addExtension('.errors')`.

This new asset is requested by the app when it fails to load a module and it prints the contents to the console so you can get compilation errors in your browser.

Partially resolves https://github.com/dart-lang/pub/issues/1618, I think we could still do something better with these errors but this is a good first step.

### CODE REVIEW

File: lib/src/dartdevc/dartdevc.dart [hunk 1]
```diff
@@ -186,6 +182,24 @@ for (let moduleName of Object.getOwnPropertyNames(modulePaths)) {
     }
 
     bootstrapContent.write('''
+// Whenever we fail to load a js module, try and request the corresponding
```
Comment (munificent) [inline lib/src/dartdevc/dartdevc.dart]: Grammar nit: "try and" -> "try to".

Also... "js" -> "JS".

File: lib/src/dartdevc/errors.dart [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:barback/barback.dart';
+
+/// An [Exception] that is thrown when the analyzer fails to create a summary.
+class AnalyzerSummaryException implements Exception {
+  /// The module that couldn't be compiled.
+  final AssetId assetId;
```
Comment (munificent) [inline lib/src/dartdevc/errors.dart]: We know it's an AssetId from the type. How about a more meaningful name. Maybe "failedModule" or just "module"?

File: lib/src/dartdevc/errors.dart [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  AnalyzerSummaryException(this.assetId, this.error);
+
+  String toString() => 'Error creating summary for module: $assetId\n\n'
+      '$error';
+}
+
+/// An [Exception] that is thrown when dartdevc compilation fails.
+class DartDevcCompilationException implements Exception {
+  /// The js module that couldn't be compiled.
```
Comment (munificent) [inline lib/src/dartdevc/errors.dart]: "js".

File: lib/src/dartdevc/errors.dart [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  DartDevcCompilationException(this.assetId, this.error);
+
+  String toString() => 'Error compiling dartdevc module: $assetId\n\n'
+      '$error';
+}
+
+/// An [Exception] that is thrown when a module is not found for an [AssetId].
+class MissingModuleException implements Exception {
+  /// The asset that a module could not be found for.
+  final AssetId assetId;
```
Comment (munificent) [inline lib/src/dartdevc/errors.dart]: There's a lot of duplication between these classes. How about a base class that contains the AssetId? Or maybe just a base class for the first two that has an AssetId and the dartdevc worker error?

File: lib/src/dartdevc/dartdevc.dart (hunk 1 above)
Comment (jakemac53) [inline lib/src/dartdevc/dartdevc.dart]: done

File: lib/src/dartdevc/errors.dart (hunk 2 above)
Comment (jakemac53) [inline lib/src/dartdevc/errors.dart]: Created base class where this in named `failedAsset`, constructor for this class takes an arg called `summaryId`.

File: lib/src/dartdevc/errors.dart (hunk 3 above)
Comment (jakemac53) [inline lib/src/dartdevc/errors.dart]: Created base class where this in named failedAsset, constructor for this class takes an arg called jsId.

File: lib/src/dartdevc/errors.dart (hunk 4 above)
Comment (jakemac53) [inline lib/src/dartdevc/errors.dart]: created a `_WorkerException` base class

**Pull request merged**